matrix/SquareMatrix.hpp

```diff
@@ -118,6 +130,119 @@ class SquareMatrix : public Matrix<Type, M, M>
         }
         return res;
     }
+
+    // zero all offdiagonal elements and keep corresponding diagonal elements
+    template <size_t width>
+    void setOffDiagZero(size_t first)
+    {
+        SquareMatrix<Type, M> &self = *this;
+        Vector<Type, width> diag_elements = self.slice<width, width>(first, first).diag();
+        self.makeDiag(first, diag_elements);
+    }
+
+    template <size_t width>
+    void makeDiag(size_t first, const Vector<Type, width> &vec)
+    {
+        SquareMatrix<Type, M> &self = *this;
+        // zero rows and columns
+        self.slice<width, M>(first, 0) = 0;
+        self.slice<M, width>(0, first) = 0;
+
+        // set diagonals
+        unsigned vec_idx = 0;
+        for (size_t idx = first; idx < first+width; idx++) {
+            self(idx,idx) = vec(vec_idx);
+            vec_idx ++;
+        }
+    }
+
+    template <size_t width>
+    void makeDiag(size_t first, Type val)
+    {
+        SquareMatrix<Type, M> &self = *this;
+        // zero rows and columns
+        self.slice<width, M>(first, 0) = 0;
+        self.slice<M, width>(0, first) = 0;
+
+        // set diagonals
+        for (size_t idx = first; idx < first+width; idx++) {
+            self(idx,idx) = val;
+        }
+    }
+
+    // make block diagnoal symmetric by taking the average of the two corresponding off diagonal values
+    template <size_t width>
+    void makeBlockSymmetric(size_t first)
+    {
+        SquareMatrix<Type, M> &self = *this;
+        if(width>1) {
+            for (size_t row_idx = first+1; row_idx < first+width; row_idx++) {
+                for (size_t col_idx = first; col_idx < row_idx; col_idx++) {
+                    Type tmp = (self(row_idx,col_idx) + self(col_idx,row_idx)) / 2;
```

Comment:
I'm not sure what kind of accuracy we are going for here, but what I've seen used before to avoid scaling issues is:
```suggestion
                    Type tmp = self(row_idx,col_idx) + (self(col_idx,row_idx) - self(row_idx,col_idx)) / 2;
```

Reply:
Do you need to do the same trick for the first summand?